highway.tex

```diff
@@ -615,9 +630,207 @@ \subsection{Checkpoints}
+
+In this section, we first outline two of possible spam patterns that creates very high amounts of messages, in order to show that without taking any preventive measures, the spam attacks can cause a resource depletion in validator nodes and hence possibly crash the whole protocol. Then, we show how using relatively simple techniques, it is possible to bound the amount of possible spam to the easily managable size. 
+
+\subsection{Classification of Basic Spam Patterns}
+
+In this subsection, we lay out two basic categories of spam attacks:
+
+In a vertical spam, the attacker does not produce any equivocations and merely produces a single long string of sequential messages. This could be done by simply setting their round exponent to an unrealistically small number. This is substantially the easier of type of spam to deal with, but the only type that can be produced without the attacker equivocating, and thus allowing them to be readily punished for it.
+
+In an equivocation bomb, instead of making many sequential messages, the attacker makes a large number of non-sequential messages. This is a somewhat trickier situation to deal with especially because as discussed above, the attacker can show different messages to different validators, and it will take some coordination between the other validators in order to determine which messages should and shouldn't be ignored.
+
+
+\subsubsection{Vertical Spam}
+
+In vertical spam, an attacker simply creates a long sequence of valid votes, faster than the other validators can readily handle. Perhaps the easiest way to deal with this issue is to set a minimum round exponent that is not too much smaller than what one would expect the fastest rate that the system can process messages. If additionally it is required that messages be timestamped and that no validator sends more than three messages in a round (measured by comparing the timestamp of the message to the timestamp of the third-most-recent message by the same validator in its history), this ensures that no attacker can create valid, non-equivocating messages at a rate much faster than that of any of the honest validators.
+
+This has one potential issue, however. Although an attacker many not be able to produce messages at a faster rate than honest validators, by delaying sending these messages, they may be able to manufacture a situation where many saved up messages are released at once. If they spent enough time saving up this sequence of messages, this may well be enough to at least temporarily shut down the system.
```

Comment:
Should we mention the "short eras" solution here?

Either that, or add the Impatience section after all?

Reply:
We probably should, although without a specific section devoted to explaining why it's a good idea anyway it would be weird to drop such an important information at the end of a subsection devoted to some exotic spam attack. ;) The section is coming though, don't worry.